### ISSUE TITLE
NullPointerException/ Unable to connect to Metabase h2 DB

### ISSUE THREAD

**Issue by mattes** [metabase/metabase#4851]
We were seeing CPU at 100% and I restarted the VM which helped before. I think that may have corrupted the H2 database. See error below. It's probably not metabase related, but maybe someone has an idea to fix this?

```
04-24 18:11:03 ERROR metabase.driver :: Failed to connect to database: org.h2.jdbc.JdbcSQLException: General error: "java.lang.NullPointerException" [50000-193]
java.lang.AssertionError: Assert failed: Unable to connect to Metabase h2 DB.
(binding [*allow-potentailly-unsafe-connections* true] (require (quote metabase.driver)) ((resolve (quote metabase.driver/can-connect-with-details?)) engine details))
        at metabase.db$verify_db_connection.invokeStatic(db.clj:320)
        at metabase.db$verify_db_connection.invoke(db.clj:313)
        at metabase.db$verify_db_connection.invokeStatic(db.clj:316)
        at metabase.db$verify_db_connection.invoke(db.clj:313)
        at metabase.db$setup_db_BANG_.invokeStatic(db.clj:368)
        at metabase.db$setup_db_BANG_.doInvoke(db.clj:362)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at metabase.core$init_BANG_.invokeStatic(core.clj:102)
        at metabase.core$init_BANG_.invoke(core.clj:81)
        at metabase.core$start_normally.invokeStatic(core.clj:186)
        at metabase.core$start_normally.invoke(core.clj:180)
        at metabase.core$_main.invokeStatic(core.clj:278)
        at metabase.core$_main.doInvoke(core.clj:273)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at metabase.core.main(Unknown Source)
04-24 18:11:03 ERROR metabase.core :: Metabase Initialization FAILED:  Assert failed: Unable to connect to Metabase h2 DB.
(binding [*allow-potentailly-unsafe-connections* true] (require (quote metabase.driver)) ((resolve (quote metabase.driver/can-connect-with-details?)) engine details))
04-24 18:11:03 INFO metabase.core :: Metabase Shutting Down ...
```
I see the following files:

```
metabase.db.mv.db  metabase.db.trace.db
```

- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
Postgres

- Metabase version: (e.x. 0.19.3)
0.23.1

- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
Linux xxxxxxxx 4.9.16-coreos-r1
Docker

- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
H2

**Comment by mattes** [comment]
I tried using the h2 recover tool, but I'm send to `/setup` after I restart metabase.
https://github.com/metabase/metabase/issues/3560#issuecomment-253632008

```
java -cp /app/metabase.jar org.h2.tools.Recover
Error: java.lang.IllegalStateException: Unsupported type 17 [1.4.193/3]
Error: java.lang.IllegalStateException: Unsupported type 17 [1.4.193/3]
```

Which creates:

```
metabase.db.h2.sql
metabase.db.mv.txt
```

```
java -cp /app/metabase.jar org.h2.tools.RunScript -script metabase.db.h2.sql -url jdbc:h2:/data/metabase.db
```

... returns without an error. But then starting metabase, it will try to setup everything from fresh.

**Comment by mattes** [comment]
Looking at the `metabase.db.h2.sql`, it seems as if the data isn't lost. So I wonder if I just need to set a flag so it's not trying to set up metabase again?

**Comment by mattes** [comment]
* The backups we had seemed to be corrupted as well.
* I didn't find out what caused the 100% CPU in the first place.
* Don't blindly restart/ force kill. Duh! 
* Using `java -cp /app/metabase.jar org.h2.tools.Recover` and running the following regex `name":"(.*?)".*query":"(([^"\\]|\\.)*)"` on the resulting `metabase.db.h2.sql` file, I was able to recover the raw queries.

* **Biggest takeaway: Don't use the default H2 database!** (for anything production related)

**Comment by tgeorge** [comment]
I ran into the same issue without new backup.

Just a bit of help if some else runs into this.  I started looking into H2 because I never dealt with it in the past.  So, I went to their website and installed the jar and copied it to the Metabase directory.  I checked out their documentation and found this their command line tools:   http://www.h2database.com/html/tutorial.html#command_line_tools

I tried their Recover function but ended up getting the java Error and it didn't make any changes to the db (the timestamp didn't change).   But then I tried the Backup recommendation `java org.h2.tools.Script -url jdbc:h2:(Your absolute Path)\metabase.db -script test.zip -options compression zip`  http://www.h2database.com/html/tutorial.html#upgrade_backup_restore and it made a zip file but I also noticed the original db's timestamp changed as well.  So launched Metabase as is and it weirdly worked.   

I am going to migrate Metabase to MySQL after this headache haha.
